decisions/websre/0005-web-sre-kubernetes-secrets-pattern.md

```diff
@@ -0,0 +1,55 @@
+# 5. web-sre-kubernetes-secrets-pattern
+
+Date: 2021-06-09
+
+## Status
+
+Accepted
+
+## Context
+
+Secrets are inconsistently managed across our various clusters.  This makes it hard to trace where secrets come from and update them.  It also increases our exposure to security issues, as our each tool used can become compromised independently.  Current known secret management includes git-crypt repos, with sops, that apply kubernetes secrets direct to the cluster; sealed secrets; and external secrets pointing at cloud provider secret storage.
+
+There are a few other methods that treat things that should be secrets incorrectly. For example, baking the 'secrets' into the docker image directly. This is a bad state to be in, and we will treat those situations with more urgency, correcting the pattern as soon as possible after this ADR is accepted.
+
+The goal of this document is to describe the happy path for passing secrets to kubernetes based applications.  In general non-kubernetes secrets management is out of scope for this document, but we will cover how serverless deployments (lambdas) and CI/CD are impacted by this choice, and what we expect those to look like after adopting the ADR.
+
+## Decision
+
+Store the secrets in the cloud provider secret manager (aws secret manager, gcp secret manager), then use external secrets (https://github.com/external-secrets/kubernetes-external-secrets) to retrieve those secrets and present them to pods as env variables.  The reason this storage method is preferred is that the cloud secret managers seem to be the best secured, most tested, easiest to use place to store secrets. External secrets connects to the secret managers and presents the secret(s) in the method most often recommended for kubernetes; the 12 Factor App pattern of using environment variables.  One other nice benefit of this system is that references to the secrets can be stored in source code, so the connection between the app and the secret storage is explicit. The secret value itself is not in source code, which has the benefit of reducing the risk of leaking the secret accidently.
+
+There are currently a few ways the secrets can be written to the cloud secret store. We have some terraform that writes the secrets directly to the secrets stores, but most secrets are populated manually by operators.  We expect both methods to be used going forward.  In general, we expect SREs/Operators with elevated privileges to manage the secrets.  This ADR intentionally doesn't include scope for devs to write new secrets, or update existing ones.
```

Comment:
What happens if secrets get lost / accidentally deleted? Are all secrets easy and quick to regenerate?

Reply:
Answered a bit more below, but a single secret being deleted is likely not too hard - aws can often recover it with a support ticket if you discover in time, any pod with the secret currently keeps running, and we created all these secrets once, so it's likely not that hard to create again.  But if all secrets get lost, it would be a big pain, for sure.

Reply:
fwiw, I agree with @duallain here, but appreciate @ajvb what you're bringing. I think the least concern for me is that AWS or GCP somehow loses all our secrets (if they do, we've probably got bigger problems in that scenario). Having seen what some of the services workflows do in using SOPS as well - ultimately ending up with sops-managed secrets decrypted on a s3 bucket to be picked up via jenkins pipelines - I have other security considerations that I would want to repair before considering that approach. Here, we're using the AWS & GCP secrets managers for what they're built for, and keeping away from embedding secrets in build pipelines, docker images, other git repos - keeping it limited ot the infrastructure fed into k8s.

Reply:
(going to respond to these comments in the below thread, since I essentially made the same comment twice)